### ISSUE TITLE
estimator built with keras.estimator.model_to_estimator fails with CollectiveAllReduceStrategy

### ISSUE THREAD

**Issue by seb1000** [tensorflow/tensorflow#27000]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
CentOS 7 on a HPC system

- TensorFlow installed from (source or binary):
binary (from anaconda)
- TensorFlow version (use command below):
1.12.0
- Python version:
3.6.8
- CUDA/cuDNN version: None
- GPU model and memory: None

**Describe the current behavior**
When converting a keras model to an estimator, it can be trained in a non-distributed environment,
but it fails with CollectiveAllReduceStrategy on multiple machines. However, when building an estimator from scratch (without keras), CollectiveAllReduceStrategy works.
**Describe the expected behavior**
estimator created from keras models should be trainable with CollectiveAllReduceStrategy.

**Code to reproduce the issue**

the following script fails when run in a distributed way in a slurm-cluster:
```

import numpy as np
import tensorflow as tf

from tensorflow.keras.datasets import mnist
import os
import sys
import json



train_number = "test"
worker_ips = sys.argv[1]
port = sys.argv[2]
worker_index = int(sys.argv[3])
modeldir = sys.argv[4]
workers = [ip+':'+port for ip in worker_ips.split(',')]

os.environ["TF_CONFIG"] = json.dumps({
    "cluster": {
        "worker": workers
    },
   "task": {"type": "worker", "index": worker_index}
})

distribution = tf.contrib.distribute.CollectiveAllReduceStrategy(
    num_gpus_per_worker=0)
config = tf.estimator.RunConfig(train_distribute=distribution, model_dir=modeldir)


batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()



x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

model = tf.keras.Sequential()
model.add(tf.keras.layers.Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(tf.keras.layers.Conv2D(64, (3, 3), activation='relu'))
model.add(tf.keras.layers.MaxPooling2D(pool_size=(2, 2)))
model.add(tf.keras.layers.Dropout(0.25))
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(128, activation='relu'))
model.add(tf.keras.layers.Dropout(0.5))
model.add(tf.keras.layers.Dense(num_classes, activation='softmax'))

model.compile(loss=tf.keras.losses.categorical_crossentropy,
              optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.001),
metrics=['accuracy'])

estimator = tf.keras.estimator.model_to_estimator(keras_model=model, config=config)

input_name = model.input_names[0]


def gen():
    for x,y in zip(x_train,y_train):
        yield x, y


def train_input_fn():
    dataset = tf.data.Dataset.from_generator(
        gen, (tf.float32, tf.float32), output_shapes=(x_train.shape[1:],y_train.shape[1:]))

    return dataset.shuffle(1).repeat(epochs).batch(batch_size)

train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn)
eval_spec = tf.estimator.EvalSpec(input_fn=train_input_fn)
tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
```

the multi-machine job is started with the following script
```
#! /bin/bash

# N must be >=2

#SBATCH -N2
#SBATCH -A snic2018-1-13
#SBATCH --time=00:30:00

nodelists=$(scontrol show hostname $SLURM_JOB_NODELIST | paste -d, -s)
IFS=',' read -r -a nodearray <<< "$nodelists"


source activate tf-dist-env

port=2222

modeldir=$(mktemp -d --tmpdir='.')
rm -r $modeldir
mkdir $modeldir

for i in $(seq 1 $SLURM_JOB_NUM_NODES); do
# convert to 0-based index
    i=$((i-1))
    echo starting job $i
    srun --nodelist="${nodearray[i]}"  -N1 -n1 python test-tf-estimators-from-keras_distributed.py ${nodelists} ${port} ${i} ${modeldir} &
done
wait
```

which produces one of the following error
```
Traceback (most recent call last):
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Upper bound check fail for input 1 from node training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter to node scoped_allocator_concat_1 input bounds = [0x2b70cc944a40, 0x2b70cc944ec0] backing_tensor bounds = [0x2b70cc01e000, 0x2b70cc4b1c28]
	 [[{{node scoped_allocator_concat_1}} = _ScopedAllocatorConcat[N=8, T=DT_FLOAT, id=1, reshape=false, sa_name="scoped_allocator_1", shape=[1199882], _device="/job:worker/replica:0/task:1/device:CPU:0"](scoped_allocator_1, training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d_1/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense_1/BiasAdd_grad/BiasAddGrad, ^training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropInput, ^training/TFOptimizer/gradients/dense/MatMul_grad/MatMul, ^training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test-tf-estimators-from-keras_distributed.py", line 100, in <module>
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/training.py", line 462, in train_and_evaluate
    estimator, train_spec, eval_spec, _TrainingExecutor)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/estimator_training.py", line 279, in train_and_evaluate
    session_config=run_config.session_config)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/distribute_coordinator.py", line 792, in run_distribute_coordinator
    task_id, session_config, rpc_layer)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/distribute_coordinator.py", line 344, in _run_single_worker
    worker_fn(strategy)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/estimator_training.py", line 246, in _worker_fn
    hooks=hooks)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 354, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1205, in _train_model
    return self._train_model_distributed(input_fn, hooks, saving_listeners)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1352, in _train_model_distributed
    saving_listeners)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1471, in _train_with_estimator_spec
    _, loss = mon_sess.run([estimator_spec.train_op, estimator_spec.loss])
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 671, in run
    run_metadata=run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1156, in run
    run_metadata=run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1255, in run
    raise six.reraise(*original_exc_info)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1240, in run
    return self._sess.run(*args, **kwargs)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1312, in run
    run_metadata=run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1076, in run
    return self._sess.run(*args, **kwargs)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Upper bound check fail for input 1 from node training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter to node scoped_allocator_concat_1 input bounds = [0x2b70cc944a40, 0x2b70cc944ec0] backing_tensor bounds = [0x2b70cc01e000, 0x2b70cc4b1c28]
	 [[{{node scoped_allocator_concat_1}} = _ScopedAllocatorConcat[N=8, T=DT_FLOAT, id=1, reshape=false, sa_name="scoped_allocator_1", shape=[1199882], _device="/job:worker/replica:0/task:1/device:CPU:0"](scoped_allocator_1, training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d_1/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense_1/BiasAdd_grad/BiasAddGrad, ^training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropInput, ^training/TFOptimizer/gradients/dense/MatMul_grad/MatMul, ^training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul)]]
2019-03-21 17:18:47.019340: W tensorflow/core/common_runtime/base_collective_executor.cc:203] BaseCollectiveExecutor::StartAbort Invalid argument: Upper bound check fail for input 1 from node training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter to node scoped_allocator_concat_1 input bounds = [0x2b7b2409eb00, 0x2b7b2409ef80] backing_tensor bounds = [0x2b7a76ae2040, 0x2b7a76f75c68]
	 [[{{node scoped_allocator_concat_1}} = _ScopedAllocatorConcat[N=8, T=DT_FLOAT, id=1, reshape=false, sa_name="scoped_allocator_1", shape=[1199882], _device="/job:worker/replica:0/task:0/device:CPU:0"](scoped_allocator_1, training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d_1/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense_1/BiasAdd_grad/BiasAddGrad, ^training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropInput, ^training/TFOptimizer/gradients/dense/MatMul_grad/MatMul, ^training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul)]]
x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples
Traceback (most recent call last):
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1334, in _do_call
    return fn(*args)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1319, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1407, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Upper bound check fail for input 1 from node training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter to node scoped_allocator_concat_1 input bounds = [0x2b7b2409eb00, 0x2b7b2409ef80] backing_tensor bounds = [0x2b7a76ae2040, 0x2b7a76f75c68]
	 [[{{node scoped_allocator_concat_1}} = _ScopedAllocatorConcat[N=8, T=DT_FLOAT, id=1, reshape=false, sa_name="scoped_allocator_1", shape=[1199882], _device="/job:worker/replica:0/task:0/device:CPU:0"](scoped_allocator_1, training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d_1/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense_1/BiasAdd_grad/BiasAddGrad, ^training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropInput, ^training/TFOptimizer/gradients/dense/MatMul_grad/MatMul, ^training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul)]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test-tf-estimators-from-keras_distributed.py", line 100, in <module>
    tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/training.py", line 462, in train_and_evaluate
    estimator, train_spec, eval_spec, _TrainingExecutor)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/estimator_training.py", line 279, in train_and_evaluate
    session_config=run_config.session_config)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/distribute_coordinator.py", line 792, in run_distribute_coordinator
    task_id, session_config, rpc_layer)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/distribute_coordinator.py", line 344, in _run_single_worker
    worker_fn(strategy)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/distribute/estimator_training.py", line 246, in _worker_fn
    hooks=hooks)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 354, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1205, in _train_model
    return self._train_model_distributed(input_fn, hooks, saving_listeners)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1352, in _train_model_distributed
    saving_listeners)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py", line 1471, in _train_with_estimator_spec
    _, loss = mon_sess.run([estimator_spec.train_op, estimator_spec.loss])
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 671, in run
    run_metadata=run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1156, in run
    run_metadata=run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1255, in run
    raise six.reraise(*original_exc_info)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1240, in run
    return self._sess.run(*args, **kwargs)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1312, in run
    run_metadata=run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1076, in run
    return self._sess.run(*args, **kwargs)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 929, in run
    run_metadata_ptr)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1152, in _run
    feed_dict_tensor, options, run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1328, in _do_run
    run_metadata)
  File "/proj/bolinc/users/x_sebsc/anaconda3/envs/tf-dist-env/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1348, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: Upper bound check fail for input 1 from node training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter to node scoped_allocator_concat_1 input bounds = [0x2b7b2409eb00, 0x2b7b2409ef80] backing_tensor bounds = [0x2b7a76ae2040, 0x2b7a76f75c68]
	 [[{{node scoped_allocator_concat_1}} = _ScopedAllocatorConcat[N=8, T=DT_FLOAT, id=1, reshape=false, sa_name="scoped_allocator_1", shape=[1199882], _device="/job:worker/replica:0/task:0/device:CPU:0"](scoped_allocator_1, training/TFOptimizer/gradients/conv2d/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropFilter, training/TFOptimizer/gradients/conv2d_1/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense/BiasAdd_grad/BiasAddGrad, training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul_1, training/TFOptimizer/gradients/dense_1/BiasAdd_grad/BiasAddGrad, ^training/TFOptimizer/gradients/conv2d_1/Conv2D_grad/Conv2DBackpropInput, ^training/TFOptimizer/gradients/dense/MatMul_grad/MatMul, ^training/TFOptimizer/gradients/dense_1/MatMul_grad/MatMul)]]
```

When instead running the following script on a single machine, it works (same as the python script for the distributed job, but without the configuration for distribution):

```

import numpy as np
import tensorflow as tf

from tensorflow.keras.datasets import mnist
import os
import sys
import json





batch_size = 128
num_classes = 10
epochs = 12

# input image dimensions
img_rows, img_cols = 28, 28

# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = mnist.load_data()



x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')

# convert class vectors to binary class matrices
y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

model = tf.keras.Sequential()
model.add(tf.keras.layers.Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(tf.keras.layers.Conv2D(64, (3, 3), activation='relu'))
model.add(tf.keras.layers.MaxPooling2D(pool_size=(2, 2)))
model.add(tf.keras.layers.Dropout(0.25))
model.add(tf.keras.layers.Flatten())
model.add(tf.keras.layers.Dense(128, activation='relu'))
model.add(tf.keras.layers.Dropout(0.5))
model.add(tf.keras.layers.Dense(num_classes, activation='softmax'))

model.compile(loss=tf.keras.losses.categorical_crossentropy,
              optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.001),
metrics=['accuracy'])

estimator = tf.keras.estimator.model_to_estimator(keras_model=model)

input_name = model.input_names[0]


def gen():
    for x,y in zip(x_train,y_train):
        yield x, y


def train_input_fn():
    dataset = tf.data.Dataset.from_generator(
        gen, (tf.float32, tf.float32), output_shapes=(x_train.shape[1:],y_train.shape[1:]))

    return dataset.shuffle(1).repeat(epochs).batch(batch_size)

train_spec = tf.estimator.TrainSpec(input_fn=train_input_fn)
eval_spec = tf.estimator.EvalSpec(input_fn=train_input_fn)
tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)
```

The following minimum example using an estimator not created with keras works with CollectiveAllReduceStrategy

```

import tensorflow as tf
import os
import json

import sys
worker_ips = sys.argv[1]
port = sys.argv[2]
worker_index = int(sys.argv[3])
modeldir = sys.argv[4]
workers = [ip+':'+port for ip in worker_ips.split(',')]

os.environ["TF_CONFIG"] = json.dumps({
    "cluster": {
        "worker": workers
    },
   "task": {"type": "worker", "index": worker_index}
})

distribution = tf.contrib.distribute.CollectiveAllReduceStrategy(
    num_gpus_per_worker=0)
config = tf.estimator.RunConfig(train_distribute=distribution, model_dir=modeldir)


def model_fn(features, labels, mode):
    layer = tf.layers.Dense(1)
    logits = layer(features)

    if mode == tf.estimator.ModeKeys.PREDICT:
        predictions = {"logits": logits}
        return tf.estimator.EstimatorSpec(mode, predictions=predictions)

    loss = tf.losses.mean_squared_error(
        labels=labels, predictions=tf.reshape(logits, []))

    if mode == tf.estimator.ModeKeys.EVAL:
        return tf.estimator.EstimatorSpec(mode, loss=loss)

    if mode == tf.estimator.ModeKeys.TRAIN:
        train_op = tf.train.GradientDescentOptimizer(0.2).minimize(loss)
        return tf.estimator.EstimatorSpec(mode, loss=loss, train_op=train_op)


def input_fn():
    print('debug: calling input_fn')
    # check whether input_fn can see info about the worker is bein
    # executed it
    # print(os.environ["TF_CONFIG"]) # this does not work...
    print(os.environ)
    print('dir():', dir())
    print('globals():', globals())
    print('locals():', locals())
    features = tf.data.Dataset.from_tensors([[1.]]).repeat(100000)
    labels = tf.data.Dataset.from_tensors(1.).repeat(100000)
    return tf.data.Dataset.zip((features, labels))


estimator = tf.estimator.Estimator(model_fn=model_fn, config=config)
train_spec = tf.estimator.TrainSpec(input_fn=input_fn, max_steps=10000)
eval_spec = tf.estimator.EvalSpec(input_fn=input_fn)
tf.estimator.train_and_evaluate(estimator, train_spec, eval_spec)

```

**Comment by seb1000** [comment]
problem solved with upgrading to 1.13.1

**Comment by seb1000** [comment]
just found out that actually the problem was not the tensorflow version, but the fact that i used 1.12.0 from conda. So it works both with 1.12.0 and 1.13.1 from pip.
